PATTERN.md

```diff
@@ -0,0 +1,283 @@
+shift and evolve as we learn new things and discover better techniques.
+
+# Practices
+
+## One cmd per repo
+
+Each repo should have precisely one main package.  Generally, its best to keep that main package in the top level
+directory so as to ease 'go get'-ing of the cmd.
+
+## main.go
+
+Try to keep to a single main.go for your cmd.  If main.go is growing larger than 100-200 lines of code, its probably indicates
+you are trying to do to much in it and should either be writing bits in sdk/ or sdk-*/ or writing pkgs in ./internal.
+
+The basic anatomy of a main.go should be:
+
+1. Setup a context
+2. Setup logging
+3. Enable optional debugging
+4. Extract configuration from environment
+5. Get a X509Source
+6. Setup any other needed things (vppagent if using it, authz policy etc)
+7. Instantiate Endpoint
+8. Create GRPC Server for endpoint
+9. Wait for exit
+
+### Setup Context
+
+Create a new context with [signalctx](https://github.com/networkservicemesh/sdk/blob/master/pkg/tools/signalctx/context.go#L32) which will be cancelled if the cmd receives OS signals like SIGTERM:
+
+```go
+ctx := signalctx.WithSignals(context.Background())
+```
+
+Optionally add a cancel function that can be used to manually cancel subsequently.
+
+```go
+ctx, cancel := context.WithCancel(ctx)
+```
+
+### Setup logging
+
+This would be the proper place and time to setup logrus Formatters, LogLevels, etc.
+Minimally though, attach a 'cmd' field to the context:
+```go
+ctx = log.WithField(ctx, "cmd", os.Args[0])
+```
+
+### Enable optional debugging
+[debug.Self()](https://github.com/networkservicemesh/sdk/blob/master/pkg/tools/debug/self.go#L53) makes it very easy to allow developers to optionally enable debugging with an Env Variable:
+```go
+if err := debug.Self(); err != nil {
+	log.Entry(ctx).Infof("%s", err)
+}
+```
+
+If the correct env variable is set, and dlv is installed, this will cause the cmd to exec dlv to debug itself.
+If there is something that prevents debugging, the error message logged will provide sufficient info to enable doing so.
+
+### Extract configuration from environment
+
+The simplest way to extract configuration from the environment is [envconfig](https://github.com/kelseyhightower/envconfig).
+It allows you to create a simple struct for your configuration, and in can unmarshal that from env variables.
+envconfig's [supported field types](https://github.com/kelseyhightower/envconfig#supported-struct-field-types) include
+anything that implements [encoding.TextUnmarshaler](https://golang.org/pkg/encoding/#TextUnmarshaler) and 
+[encoding.BinaryUnmarshaler](https://golang.org/pkg/encoding/#BinaryUnmarshaler).  This allows native use of handy things
+like url.URL and net.IP.  envconfig also interprets arrays as meaning that the env variable is a comma delimited list
+of the type of the array.
+
+[Example Config struct](https://github.com/networkservicemesh/cmd-forwarder-vppagent/blob/master/main.go#L46):
+```go
+type Config struct {
```

Comment:
Individual applications could have this changes, so best way is to make it empty? Thoughts